### PR TITLE
add context for unit and e2e testing

### PR DESCRIPTION
**Pull request by jeffvg**
#### Changes proposed in this Pull Request:
- update for unit testing
- update for e2e testing
- Needs some serious JB wordsmithing

### CODE REVIEW

File: src/docs/development/testing.mdx [hunk 1]
```diff
@@ -8,12 +8,50 @@ Tests come in a variety of different flavors. The following sections discuss Uni
...
+
+Example component with a button.
+
+- We should test what the user views i.e. rendering and what the user can do i.e. interactions.
+
+- If there is visual change after clicking , test for it.
+
+- If there is a value returned after clicking, test for it.
+
+```sh
```
Comment (huayunh) [inline src/docs/development/testing.mdx:41]:
```suggestion
```tsx
```

File: src/docs/development/testing.mdx [hunk 2]
```diff
@@ -1,33 +1,75 @@
...
+
+<Divider />
+
+# Unit Tests
+
+Unit tests target the smallest pieces of your application (such as verifying inputs and outputs of a class method or function). These will likely make up the bulk of your tests because they are quick to write, easy to automate, and can help you achieve significant code coverage.
+
+> Code Coverage measures the percentage of your code that is verified by your tests.
+
+A unit test shouldn't have any external dependencies, such as other methods or APIs. By keeping unit tests isolated in this manner, it's easy to identify the cause of a failing test an implement a fix quickly. When you start combining features and methods together, you move into the realm of _integration testing_.
```
Comment (joebochill) [inline src/docs/development/testing.mdx:29]:
```suggestion
A unit test shouldn't have any external dependencies, such as other methods or APIs. By keeping unit tests isolated in this manner, it's easy to identify the cause of a failing test and implement a fix quickly. When you start combining features and methods together, you move into the realm of _integration testing_.
```

File: src/docs/development/testing.mdx [hunk 3]
```diff
@@ -1,33 +1,75 @@
...
 
-Angular, React, and React Native all come with built-in unit testing frameworks (if you are using the respective CLI tools). Ionic can be unit tested using the same tools you would use for a standard Angular application.
+For example, consider testing an Open Button that should open a modal dialog when clicked. Your test should check that a button with the _Open_ label is rendered on the screen, and when it is clicked, the dialog is rendered on the screen with the correct content. You may be tempted to identify the button by a class or id, or test that a particular function is called when the button is pressed - these tests are not as effective because they are more likely to break if the implementation of the component changes, even though the end behavior is the same.
 
-For consistency, you should stick to using these standard tools unless you have a good reason for using alternatives.
+In general:
+
+-   Test what the user sees (i.e., rendered output) and what the user can do (i.e., interactions).
```
Comment (joebochill) [inline src/docs/development/testing.mdx:49]:
```suggestion
-   Test what the user sees (i.e., rendered output) and what the user can do  (i.e., interactions).
```

File: src/docs/development/testing.mdx [hunk 4]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+End-to-end tests rely on the ability to find elements on the screen and interact with them in an automated way. This means that you need a reliable mechanism for selecting elements. Depending on your testing framework, there are different ways to select elements, including:
+
+-   **CSS Selectors**: find elements by using CSS classes or rules (e.g., `.your-class-name`)
+-   **Element Selectors**: find elements inside the elements (e.g., `#\@\@pxb-drawerlayout-content > div > header > div > button`)
+-   **xPath Selectors**: find selector within xPath (e.g., `//*[@id="@@pxb-drawerlayout-content"]/div/header/div/button`)
+-   **ID Selectors**: find attribute id within elements (e.g., `#login-button`)
+
+See selector strategies on [browserstack](https://www.browserstack.com/guide/css-selectors-in-selenium) and [software testing help](https://www.softwaretestinghelp.com/css-selector-selenium-locator-selenium-tutorial-6/)
```
Comment (joebochill) [inline src/docs/development/testing.mdx:117]:
```suggestion
> You can read more about different selector strategies on [browserstack](https://www.browserstack.com/guide/css-selectors-in-selenium) and [software testing help](https://www.softwaretestinghelp.com/css-selector-selenium-locator-selenium-tutorial-6/).
```

File: src/docs/development/testing.mdx [hunk 5]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+```html
+<button test-id="login-button" class="default-button">Login</button>
+```
+
+Selector to use in a test
+
+```
+WebElement click = driver.findElement(By.id("login-button"));
+```
```
Comment (joebochill) [inline src/docs/development/testing.mdx:131]: Does by.id work here if the attribute is test-id?

Comment (joebochill) [reply src/docs/development/testing.mdx:131]:
```suggestion
```html
// Component code
<button test-id="login-button" class="default-button">Login</button>

// Selector to use in a test
WebElement click = driver.findElement(By.id("login-button"));
```
```

File: src/docs/development/testing.mdx [hunk 6]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+## Testing Frameworks
+
+There are many tools available for configuring automated E2E tests. Some of the more popular tools include:
+
+-   Selenium
+-   Cypress
+-   WebDriverIO
```
Comment (joebochill) [inline src/docs/development/testing.mdx:139]: Do these have a main documentation page that we could link to here?

File: src/docs/development/testing.mdx [hunk 7]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+### Selenium
+
+Selenium is a popular automation testing suite which can be used to automate the desktop and mobile web browser interactions. You can perform automation testing by writing code in any of your preferred language supported by Selenium.
```
Comment (joebochill) [inline src/docs/development/testing.mdx:143]:
```suggestion
Selenium is a popular automation testing suite which can be used to automate the desktop and mobile web browser interactions. You can write test code in any of the languages supported by Selenium.
```

File: src/docs/development/testing.mdx [hunk 8]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+Selenium test suite has several test frameworks available and each one can be custmized to your project needs.
```
Comment (joebochill) [inline src/docs/development/testing.mdx:145]:
```suggestion
The Selenium test suite has several test frameworks available and each one can be customized to your project needs.
```

File: src/docs/development/testing.mdx [hunk 9]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+For more information see [browserstack](https://www.browserstack.com/guide/selenium-framework) selenium guide or [selenium on github](https://github.com/SeleniumHQ/selenium) documentation.
```
Comment (joebochill) [inline src/docs/development/testing.mdx:147]:
```suggestion
For more information, check out the browserstack [selenium guide](https://www.browserstack.com/guide/selenium-framework) or the official [selenium documentation](https://github.com/SeleniumHQ/selenium.
```

File: src/docs/development/testing.mdx [hunk 10]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+For selenium tutorial series see [software testing help](https://www.softwaretestinghelp.com/selenium-tutorial-1/)
```
Comment (joebochill) [inline src/docs/development/testing.mdx:149]:
```suggestion
There are also a number of [Selenium tutorials](https://www.softwaretestinghelp.com/selenium-tutorial-1/) available online to help you get started.
```

File: src/docs/development/testing.mdx [hunk 11]
```diff
@@ -38,24 +80,90 @@ yarn test
...
 ### Cypress
 
-Cypress is an open source testing framework that boasts support for any modern JavaScript framework. It works well for E2E tests in both Angular and React applications and is easier to use than some of the more traditional tools, such as Selenium.
+Cypress is a popular open source testing framework that boasts support for any modern JavaScript framework. It works well for E2E tests in both Angular and React applications and is easier to use than some of the more traditional tools, such as Selenium.
 
 For a more detailed write-up on Cypress, check out the official [Cypress Documentation](https://docs.cypress.io/guides/getting-started/installing-cypress.html#System-requirements).
+
+### WebDriverIO
+
+WebdriverIO is a popular Javascript based test automation framework built over nodeJs. It is an open-source project developed for the automation testing community. WebdriverIO is extendible, compatible and feature-rich.
```
Comment (joebochill) [inline src/docs/development/testing.mdx:159]:
```suggestion
WebdriverIO is a popular Javascript based test automation framework built on top of node.js. It is an open-source project developed for the automation testing community. WebdriverIO is extendible, compatible and feature-rich.
```

File: src/docs/development/testing.mdx [hunk 12]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+For more information see [webdriver.io](https://webdriver.io/) and [webdriver.io](https://webdriver.io/docs/gettingstarted.html) getting started.
```
Comment (joebochill) [inline src/docs/development/testing.mdx:161]:
```suggestion
For more information, check out the official [Webdriver.io documentation](https://webdriver.io/) and their [getting started guide](https://webdriver.io/docs/gettingstarted.html).
```

File: src/docs/development/testing.mdx [hunk 13]
```diff
@@ -38,24 +80,90 @@ yarn test
...
+
+<Divider />
+
+# Mobile Device Testing
+
+Coming soon...
```
Comment (joebochill) [inline src/docs/development/testing.mdx:167]: Should we add something here to talk about the different options that are available (at a high level)? Or just leave this our for now until later?

**Review by joebochill — CHANGES_REQUESTED** [review]

**Comment by joebochill** [comment]
also, do a `yarn precommit` to update the search-results files with the new content.

File: src/docs/development/testing.mdx [hunk 14]
```diff
@@ -44,8 +44,6 @@ When unit testing components, the most effective tests are based on what is actu
 
 For example, consider testing an Open Button that should open a modal dialog when clicked. Your test should check that a button with the _Open_ label is rendered on the screen, and when it is clicked, the dialog is rendered on the screen with the correct content. You may be tempted to identify the button by a class or id, or test that a particular function is called when the button is pressed - these tests are not as effective because they are more likely to break if the implementation of the component changes, even though the end behavior is the same.
 
-In general:
-
 -   Test what the user sees (i.e., rendered output) and what the user can do (i.e., interactions).
```
Comment (joebochill) [inline src/docs/development/testing.mdx:47]: missing a space here
```suggestion
-   Test what the user sees (i.e., rendered output) and what the user can do  (i.e., interactions).
```

File: src/docs/development/testing.mdx [hunk 15]
```diff
@@ -114,39 +110,37 @@ End-to-end tests rely on the ability to find elements on the screen and interact
...
-<button test-id="login-button" class="default-button">Login</button>
-```
+// Component code
 
-Selector to use in a test
+<button id="login-button" class="default-button">Login</button>
 
-```
-WebElement click = driver.findElement(By.id("login-button"));
+// Selector to use in a test WebElement click = driver.findElement(By.id("login-button"));
```
Comment (joebochill) [inline src/docs/development/testing.mdx:124]: looks like prettier may have messed this code block up

Comment (jeffvg) [reply src/docs/development/testing.mdx]: think its better now

**Review by joebochill — APPROVED** [review]

**Pull request merged**